File: sample/objective-c/GNWrapperAdSampleApp/GNWrapperAdSampleApp/ViewController.m
//
//  ViewController.m
//  GNWrapperAdSampleApp
//

#import "ViewController.h"
#import "Util.h"
@import Firebase;
@import GoogleMobileAds;
#import <GNWrapperAdSdk/GNWrapperAdSDK.h>
#import <PrebidMobile/PrebidMobile-Swift.h>

// Ad Size.
#define ADMANAGER_AD_SIZE kGADAdSizeBanner
//#define ADMANAGER_AD_SIZE kGADAdSizeMediumRectangle

@interface ViewController () <GADBannerViewDelegate, GADAppEventDelegate, GNWrapperAdBannerDelegate>

@property(nonatomic, weak) FIRRemoteConfig *remoteConfig;
@property(nonatomic, retain) GNWrapperAdBanner *gnWrapperAd;
@property(nonatomic, retain) GNCustomTargetingParams *targetParams;
@property(nonatomic, strong) DFPBannerView *bannerView;

@end

@implementation ViewController

// For Firebase.
static NSString* const FIREBASE_DEFAULT_REMOTE_CONFIG = @"DefaultRemoteConfig";
static double const FIREBASE_FETCH_TIME_INTERVAL_SECONDS = 720;
static NSString* const FIREBASE_KEY_CONFIG = @"GNWrapperConfig_iOS";
//static NSString* const FIREBASE_KEY_CONFIG = @"GNWrapperConfig_iOS_Prebid";
//static NSString* const FIREBASE_KEY_CONFIG = @"GNWrapperConfig_iOS_Pubmatic";
// For GNWrapperSDK.
static const bool GNWRAPPERSDK_TEST_MODE = true;
// For AdManager.
static const bool ADMANAGER_DEVELOPER_MODE = true;


- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
        [Util checkIdfa];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self initFunction];
            [self requestTargetingData];
        });
    });
}

- (void)initFunction {
    _remoteConfig = [FIRRemoteConfig remoteConfig];
    FIRRemoteConfigSettings *remoteConfigSettings = [[FIRRemoteConfigSettings alloc] init];
    remoteConfigSettings.minimumFetchInterval = FIREBASE_FETCH_TIME_INTERVAL_SECONDS;
    _remoteConfig.configSettings = remoteConfigSettings;
    [_remoteConfig setDefaultsFromPlistFileName:FIREBASE_DEFAULT_REMOTE_CONFIG];

    [GNWrapperAdSDK setLogPriority:GNLogPriorityInfo];
    [GNWrapperAdSDK setTestMode:GNWRAPPERSDK_TEST_MODE];
    _gnWrapperAd = [[GNWrapperAdBanner alloc] init];
    [_gnWrapperAd setHidden:true];
    [self.view addSubview:_gnWrapperAd];
    _gnWrapperAd.frame = CGRectMake(0, 0, ADMANAGER_AD_SIZE.size.width, ADMANAGER_AD_SIZE.size.height);
    _gnWrapperAd.center = CGPointMake(self.view.center.x, self.view.center.y);

    _bannerView = [[DFPBannerView alloc] initWithAdSize:ADMANAGER_AD_SIZE];
    _bannerView.delegate = self;
    _bannerView.appEventDelegate = self;
    _bannerView.rootViewController = self;
    [self.view addSubview:_bannerView];
    _bannerView.center = CGPointMake(self.view.center.x, self.view.center.y);
}

- (void)requestTargetingData {
    [_remoteConfig fetchWithCompletionHandler:^(FIRRemoteConfigFetchStatus status, NSError *error) {
        if (status == FIRRemoteConfigFetchStatusSuccess) {
            [self.remoteConfig activateWithCompletion:^(BOOL changed, NSError * _Nullable error) {
                NSString *data = self.remoteConfig[FIREBASE_KEY_CONFIG].stringValue;
                NSLog(@"ViewController: Get data = %@", data);

                [self.gnWrapperAd initWithLoad:data delegate:self viewController:self];
            }];
        } else {
            NSString *data = self.remoteConfig[FIREBASE_KEY_CONFIG].stringValue;
            NSLog(@"ViewController: Not fetched = %@", data);

            [self.gnWrapperAd initWithLoad:data delegate:self viewController:self];
        }
    }];
}

- (void)requestAdManagerBanner {
    [_gnWrapperAd setHidden:true];
    [_bannerView setHidden:false];
    _bannerView.adUnitID = _targetParams.unitId;

    DFPRequest * request = [DFPRequest request];
    request.customTargeting = _targetParams.targetParams;
    if (ADMANAGER_DEVELOPER_MODE) {
        request.testDevices = @[[Util admobDeviceID]];
    }
    [self.bannerView loadRequest:request];
    NSLog(@"ViewController: requestAdManagerBanner");
}

- (void)reqestBanner {
    if ([_gnWrapperAd isNecessaryShow]) {
        [_gnWrapperAd setHidden:false];
        [_bannerView setHidden:true];

        [_gnWrapperAd show];
        [_gnWrapperAd requestRefresh:_gnWrapperAd];
    }
}


// GNWrapperAdBannerDelegate

- (void)onComplete:(GNWrapperAdBanner*)view params:(GNCustomTargetingParams *)params {
    _targetParams = params;
    NSLog(@"ViewController: onComplete unitid = %@",params.unitId);
    for (NSString* key in [params.targetParams allKeys]) {
        NSLog(@"ViewController: onComplete key = %@ , val = %@", key, [params.targetParams objectForKey:key]);
    }

    [self requestAdManagerBanner];
}

- (void)onError:(GNWrapperAdBanner*)view error:(NSError *)error{
    NSLog(@"ViewController: onError = %@", error.localizedDescription);
}


// GADBannerViewDelegate

/// Tells the delegate an ad request loaded an ad.
- (void)adViewDidReceiveAd:(DFPBannerView *)adView {
    NSLog(@"ViewController: adViewDidReceiveAd");
    [_gnWrapperAd requestRefresh:_bannerView];
    if ([@"Prebid" isEqualToString:[_gnWrapperAd getAfterLoadedGAMBanner]]) {
        [AdViewUtils findPrebidCreativeSize:adView
                                    success:^(CGSize size) {
                                        [adView resize:GADAdSizeFromCGSize(size)];
                                    } failure:^(NSError * _Nonnull error) {
                                        [GNLog logWithPriority:GNLogPriorityError message:[NSString stringWithFormat:@"ViewController: adViewDidReceiveAd error = %@", [error localizedDescription]]];
                                    }];
    }
}

/// Tells the delegate an ad request failed.
- (void)adView:(DFPBannerView *)adView
    didFailToReceiveAdWithError:(GADRequestError *)error {
    NSLog(@"ViewController: didFailToReceiveAdWithError: %@", [error localizedDescription]);
}


/// GADAppEventDelegate.

// Called when the banner receives an app event.
- (void)adView:(DFPBannerView *)banner didReceiveAppEvent:(NSString *)name withInfo:(NSString *)info {
    NSLog(@"ViewController: didReceiveAppEvent: %@", name);
    
    if ([@"pubmaticdm" isEqualToString:name]) {
        NSLog(@"GNHBPubmaticAdapter delegate");
        [self reqestBanner];
    }
}

@end

File: sample/swift/GNWrapperAdSampleApp/GNWrapperAdSampleApp/ViewController.swift
//
//  ViewController.swift
//  GNWrapperAdSampleApp
//

import UIKit
import Firebase
import GoogleMobileAds

class ViewController: UIViewController {

    // For Firebase.
    let FIREBASE_DEFAULT_REMOTE_CONFIG: String = "DefaultRemoteConfig"
    let FIREBASE_FETCH_TIME_INTERVAL_SECONDS: Double = 720
    let FIREBASE_KEY_CONFIG: String = "GNWrapperConfig_iOS"
//    let FIREBASE_KEY_CONFIG: String = "GNWrapperConfig_iOS_Prebid"
//    let FIREBASE_KEY_CONFIG: String = "GNWrapperConfig_iOS_Pubmatic"
    // For GNWrapperSDK.
    let GNWRAPPERSDK_TEST_MODE: Bool = true
    // For AdManager.
    let ADMANAGER_AD_SIZE: GADAdSize = kGADAdSizeBanner
//    let ADMANAGER_AD_SIZE: GADAdSize = kGADAdSizeMediumRectangle
    let ADMANAGER_DEVELOPER_MODE: Bool = true

    var remoteConfig: RemoteConfig! = nil
    var gnWrapperAd: GNWrapperAdBanner! = nil
    var targetParams: GNCustomTargetingParams! = nil
    var bannerView: DFPBannerView! = nil

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        DispatchQueue.global().async {
            Util.checkIdfa()
            DispatchQueue.main.async {
                self.initFunction()
                self.requestTargetingData()
            }
        }
    }

    func initFunction() {
        self.remoteConfig = RemoteConfig.remoteConfig()
        self.remoteConfig.setDefaults(fromPlist: FIREBASE_DEFAULT_REMOTE_CONFIG)

        GNWrapperAdSDK.setLogPriority(GNLogPriorityInfo)
        GNWrapperAdSDK.setTestMode(GNWRAPPERSDK_TEST_MODE)
        self.gnWrapperAd = GNWrapperAdBanner.init()
        self.gnWrapperAd.isHidden = true
        self.view.addSubview(self.gnWrapperAd)
        self.gnWrapperAd.frame = CGRect(x: 0, y: 0, width: ADMANAGER_AD_SIZE.size.width, height: ADMANAGER_AD_SIZE.size.height)
        self.gnWrapperAd.center = CGPoint(x: self.view.center.x, y: self.view.center.y)

        self.bannerView = DFPBannerView.init(adSize: ADMANAGER_AD_SIZE)
        self.bannerView.delegate = self
        self.bannerView.appEventDelegate = self
        self.bannerView.rootViewController = self
        self.view.addSubview(self.bannerView)
        self.bannerView.center = CGPoint(x: self.view.center.x, y: self.view.center.y)
    }

    func requestTargetingData() {
        remoteConfig.fetch(withExpirationDuration: TimeInterval(FIREBASE_FETCH_TIME_INTERVAL_SECONDS)) { (status, error) -> Void in
            if (status == .success) {
                self.remoteConfig.activate() { (changed, error) in
                    let data = self.remoteConfig[self.FIREBASE_KEY_CONFIG].stringValue
                    print("ViewController: Get data = \(data!)")

                    self.gnWrapperAd.initWithLoad(data, delegate: self, viewController: self)
                }
            } else {
                let data = self.remoteConfig[self.FIREBASE_KEY_CONFIG].stringValue
                print("ViewController: Not fetched = \(data!)")

                self.gnWrapperAd.initWithLoad(data, delegate: self, viewController: self)
            }
        }
    }

    func requestAdManagerBanner() {
        self.gnWrapperAd.isHidden = true
        self.bannerView.isHidden = false
        self.bannerView.adUnitID = self.targetParams.unitId

        let request: DFPRequest = DFPRequest.init()
        request.customTargeting = (self.targetParams.targetParams as! [AnyHashable : Any])
        if (ADMANAGER_DEVELOPER_MODE) {
            request.testDevices = [Util.admobDeviceID()]
        }
        self.bannerView.load(request)
        print("ViewController: requestAdManagerBanner")
    }

    func reqestBanner() {
        if (self.gnWrapperAd.isNecessaryShow()) {
            self.gnWrapperAd.isHidden = false
            self.bannerView.isHidden = true

            self.gnWrapperAd.show()
            self.gnWrapperAd.requestRefresh(self.gnWrapperAd)
        }
    }

}
// GNWrapperAdBannerDelegate
extension ViewController: GNWrapperAdBannerDelegate {

    func onComplete(_ view:GNWrapperAdBanner, params: GNCustomTargetingParams) {
        self.targetParams = params
        print("ViewController: onComplete unitid = \(params.unitId!)")
        for key in params.targetParams.allKeys {
            print("ViewController: onComplete key = \(key) , val = \(params.targetParams.object(forKey: key)!)")
        }

        requestAdManagerBanner()
    }

    func onError(_ view:GNWrapperAdBanner) {
        print("ViewController: onError")
    }

}
extension ViewController: GADBannerViewDelegate {

    /// Tells the delegate an ad request loaded an ad.
    func adViewDidReceiveAd(_ bannerView: GADBannerView) {
        print("ViewController: adViewDidReceiveAd")
        self.gnWrapperAd.requestRefresh(self.bannerView)
        if ("Prebid" == self.gnWrapperAd.getAfterLoadedGAMBanner()) {
            AdViewUtils.findPrebidCreativeSize(self.bannerView,
                success: { (size) in
                    self.bannerView.resize(GADAdSizeFromCGSize(size))
                },
                failure: { (error) in
                    print("ViewController: adViewDidReceiveAd error = \(error.localizedDescription)")
                }
            )
        }
    }

    /// Tells the delegate an ad request failed.
    func adView(_ bannerView: GADBannerView, didFailToReceiveAdWithError error: GADRequestError) {
        print("ViewController: didFailToReceiveAdWithError: \(error.localizedDescription)")
    }
    
}
extension ViewController: GADAppEventDelegate {

    // Called when the banner receives an app event.
    func adView(_ banner: GADBannerView, didReceiveAppEvent name: String, withInfo info: String?) {
        print("ViewController: didReceiveAppEvent: \(name)")

        if ("pubmaticdm" == name) {
            print("GNHBPubmaticAdapter delegate")
            reqestBanner()
        }
    }

}

